### PR TITLE
osd/OSD: choose heartbeat peers more carefully 

### PR DESCRIPTION
**Pull request by xiexingguo**
By default, monitor requires at least two valid failure votes/reports from
different hosts to mark an OSD down, which turns out to be impossible sometimes
for a replicated-pool of size of 2 in those clusters made up of hosts
with contiguous labeled OSDs.

This patch instead does a breadth-first search based on the highest level of
failure domain at cluster-wide, to try to make heartbeat peers can cover all failure domains
whenever possible, which can hopefully help accelerating osd failure detection
in the above case..

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

**Comment by xiexingguo** [comment]
Jenkins render docs

**Comment by xiexingguo** [comment]
@liewegas Can you look? Thanks!

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -5187,3 +5187,146 @@ int OSDMap::parse_osd_id_list(const vector<string>& ls, set<int> *out,
...
+  auto it = loc.find(crush->get_type_name(1));
+  if (it != loc.end())
+    my_host_name = it->second;
+  string my_failure_domain_name = "unknown";
+  it = loc.find(crush->get_type_name(failure_domain));
+  if (it != loc.end())
+    my_failure_domain_name = it->second;
+
+  skip.insert(n); // skip myself
+  set<int> osds;
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: not used

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -5187,3 +5187,146 @@ int OSDMap::parse_osd_id_list(const vector<string>& ls, set<int> *out,
...
+  set<int> raw;
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: not needed.. just use the below loop to build raw in place of the loop over raw?

File: src/osd/OSDMap.cc [hunk 3]
```diff
@@ -5187,3 +5187,146 @@ int OSDMap::parse_osd_id_list(const vector<string>& ls, set<int> *out,
...
+        if (o.empty())
+          continue;
+        auto it = o.begin();
+        l.push_back(*it);
+        o.erase(it);
+        --num;
+      }
+    }
+  }
+  list<int> candidates;
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: vector

**Comment by liewegas** [comment]
My main concern is just the complexity of the function in the case where there are lots of osds (10k+)... lots of stl maps of strings.

Also, I think we need to keep the +1/-1 heartbeat peers to ensure that we have a fully connected graph (instead of just relying on a high probability of randomly constructing one).

**Comment by xiexingguo** [comment]
Comments addressed.

> My main concern is just the complexity of the function in the case where there are lots of osds (10k+)... lots of stl maps of strings.

Agree. I add an option to conditionally switch to the legacy behaviour once the number of osds is greater than 200 (by default) . Is that make sense to you @liewegas ?

**Comment by xiexingguo** [comment]
retest this please

**Comment by liewegas** [comment]
That could work. I wonder if an alternative approach might also makes sense:

start with this osd.
get the direct ancestor.
pick one child of that ancestor at random that is not us.  (or, alternatively, pick the 'next' child that isn't us.)  if it's not an osd/leaf, randomly descend to pick an osd.
repeat (proceed to the next immediate ancestor).

I think the complexity of something like that is more like O(log n), assuming bucket sizes are bounded?  It's more or less what the n+1 and n-1 code does not, but repeated at least level of the hierarchy.

**Comment by xiexingguo** [comment]
@liewegas I end up fixing this problem in a slightly different but more efficient way, see https://github.com/ceph/ceph/pull/23487

**Pull request closed without merging**